src/server/conn.rs

```diff
@@ -0,0 +1,180 @@
+use crate::{rt::tokio_executor::TokioExecutor, util::PrependAsyncRead};
+use http::{Request, Response};
+use http_body::Body;
+use hyper::{
+    body::Incoming,
+    server::conn::{http1, http2},
+    service::Service,
+};
+use std::{error::Error as StdError, io::Cursor, marker::Unpin};
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite};
+
+const H2_PREFACE: &[u8] = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+
+/// Http1 or Http2 connection.
+pub struct AutoConnection {
+    h1: http1::Builder,
+    h2: http2::Builder<TokioExecutor>,
+}
+
+impl AutoConnection {
+    /// Create a new AutoConn.
+    pub fn new() -> Self {
+        Self {
+            h1: http1::Builder::new(),
+            h2: http2::Builder::new(TokioExecutor::new()),
+        }
+    }
+
+    /// Http1 configuration.
+    pub fn h1(&mut self) -> &mut http1::Builder {
+        &mut self.h1
+    }
+
+    /// Http2 configuration.
+    pub fn h2(&mut self) -> &mut http2::Builder<TokioExecutor> {
+        &mut self.h2
+    }
+
+    /// Bind a connection together with a [`Service`].
+    pub async fn serve_connection<I, S, B>(&self, mut io: I, service: S) -> crate::Result<()>
+    where
+        S: Service<Request<Incoming>, Response = Response<B>> + Send,
+        S::Future: Send + 'static,
+        S::Error: Into<Box<dyn StdError + Send + Sync>>,
+        B: Body + Send + 'static,
+        B::Data: Send,
+        B::Error: Into<Box<dyn StdError + Send + Sync>>,
+        I: AsyncRead + AsyncWrite + Unpin + 'static,
+    {
+        let mut buf = Vec::new();
+
+        let protocol = loop {
+            if buf.len() < 24 {
+                io.read_buf(&mut buf).await?;
+
+                let len = buf.len().min(H2_PREFACE.len());
+
+                if buf[0..len] != H2_PREFACE[0..len] {
+                    break Protocol::H1;
+                }
+            } else {
+                break Protocol::H2;
+            }
+        };
+
+        let io = PrependAsyncRead::new(Cursor::new(buf), io);
```

Comment:
There was a similar-ish thing in `hyper::common::io::Rewind`, I don't know if it helps to compare. We can always tweak the internals to optimize, though.

Reply:
That one looks shorter. We should be able to reuse it.